src/i2c.rs

```diff
@@ -773,6 +810,353 @@ impl<'p> I2CController<'p> {
+            });
+            self.regs.smbn_st().write(|w| w.nmatch().set_bit());
+
+            let st = loop {
+                let (st, fif_cts) = self
+                    .wait_for(|| {
+                        let st = self.regs.smbn_st().read();
+                        let rxf_sts = self.regs.smbn_rxf_sts().read();
+                        let fif_cts = self.regs.smbn_fif_cts().read();
+                        if st.ber().bit_is_set()
+                            || st.slvstp().bit_is_set()
+                            || st.sdast().bit_is_set()
+                            || rxf_sts.rx_thst().bit_is_set()
+                            || fif_cts.slvrstr().bit_is_set()
+                        {
+                            Some((st, fif_cts))
+                        } else {
+                            None
+                        }
+                    })
+                    .await;
+
+                if st.ber().bit_is_set() || st.slvstp().bit_is_set() || fif_cts.slvrstr().bit_is_set() {
+                    break st;
+                }
+
+                for b in buffer.iter_mut() {
+                    *b = self.regs.smbn_sda().read().bits();
+                }
+
+                self.regs.smbn_rxf_sts().write(|w| w.rx_thst().set_bit());
+
+                handler(addr, ListenCommand::PartialWrite(&buffer));
+            };
+
+            if st.ber().bit_is_set() {
+                handler(0, ListenCommand::BusError);
+            }
+
+            // Read remaining bytes
+            let rest = self.regs.smbn_rxf_sts().read().rx_bytes().bits();
+            if rest > GROUP_SIZE {
+                for b in buffer.iter_mut() {
+                    *b = self.regs.smbn_sda().read().bits();
+                }
+
+                handler(addr, ListenCommand::WriteFinished(&buffer));
+            }
+
+            for b in buffer.iter_mut().take(rest.into()) {
+                *b = self.regs.smbn_sda().read().bits();
+            }
+
+            handler(addr, ListenCommand::WriteFinished(&buffer[..rest.into()]));
+
+            // Reset and prepare for next transaction
+            // Note that we leave slvstp for the caller to deal with
+            self.regs.smbn_rxf_ctl().write(|w| w.thr_rxie().clear_bit());
+            self.regs
+                .smbn_fif_cts()
+                .write(|w| w.clr_fifo().set_bit().slvrstr().set_bit());
+            // Clearing the Ber bit is enough to handle the bus error
+            self.regs.smbn_st().write(|w| w.ber().set_bit());
+        }
+    }
+
+    /// Listen for i2c interactions targeting the specified addresses. The handler will be called
+    /// to handle the various transactions. The listening can be stopped by setting the provided
+    /// watched value to true
+    pub async fn listen<const N: usize>(
+        &mut self,
+        addresses: &[u8],
+        mut handler: impl FnMut(u8, ListenCommand),
+        stop: &mut watch::Receiver<'_, impl RawMutex, bool, N>,
```

Comment:
I think this could use some further abstraction. Maybe wrap this in a `CancellationToken` struct? That struct then has one function for the user named `cancel` or `stop` or whatever. That would simplify the interface.

Reply:
Or maybe cancel when dropping the future?